src/psyclone/psyGen.py

```diff
@@ -5356,13 +5538,112 @@ def _if_stmt_handler(self, node, parent):
         :returns: PSyIR representation of node
         :rtype: :py:class:`psyclone.psyGen.IfBlock`
         '''
-        ifblock = IfBlock(parent=parent)
+        ifblock = IfBlock(parent=parent, annotation='was_single_stmt')
         self.process_nodes(parent=ifblock, nodes=[node.items[0]],
                            nodes_parent=node)
-        self.process_nodes(parent=ifblock, nodes=[node.items[1]],
+        ifbody = Schedule(parent=ifblock)
+        ifblock.addchild(ifbody)
+        self.process_nodes(parent=ifbody, nodes=[node.items[1]],
                            nodes_parent=node)
         return ifblock
 
+    def _case_construct_handler(self, node, parent):
+        '''
+        Transforms an fparser2 Case_Construct to the PSyIR representation.
+
+        :param node: node in fparser2 tree.
+        :type node: :py:class:`fparser.two.Fortran2003.Case_Construct`
+        :param parent: Parent node of the PSyIR node we are constructing.
+        :type parent: :py:class:`psyclone.psyGen.Node`
+        :returns: PSyIR representation of node
+        :rtype: :py:class:`psyclone.psyGen.IfBlock`
+        :raises InternalError: If the fparser2 tree has an unexpected \
+            structure.
+        :raises NotImplementedError: If the fparser2 tree contains an \
+            unsupported structure and should be placed in a CodeBlock.
+        '''
+        from fparser.two import Fortran2003
+        # Check that the fparser2 parsetree has the expected structure
+        if not isinstance(node.content[0], Fortran2003.Select_Case_Stmt):
+            raise InternalError(
+                "Failed to find opening case statement in: "
+                "{0}".format(str(node)))
+        if not isinstance(node.content[-1], Fortran2003.End_Select_Stmt):
+            raise InternalError(
+                "Failed to find closing case statement in: "
+                "{0}".format(str(node)))
+
+        # Search for all the CASE clauses in the Case_Construct
+        clause_indices = []
+        selector = None
+        for idx, child in enumerate(node.content):
+            child._parent = node  # Retrofit parent info
+            if isinstance(child, Fortran2003.Select_Case_Stmt):
+                selector = child.items[0]
+            if isinstance(child, (Fortran2003.Case_Stmt,
+                                  Fortran2003.End_Select_Stmt)):
+                clause_indices.append(idx)
+                # Case Default and value Ranges not supported yet
+                if 'DEFAULT' in str(child):
+                    raise NotImplementedError("Case Default Statement")
+                if ':' in str(child):
+                    raise NotImplementedError("Case Value Range Statement")
+
```

Comment:
Please could you add an `else` here and raise an InternalError that we've found a child of a Select_Case_Stmt that we weren't expecting.

Reply:
`else` is valid case, I reordered the code with the `clause_indices.append(idx)` inside the else to make this easier to read.

Reply:
Thanks although what I meant was we should protect ourselves against the `child` *not* being an instance of either `Case_Stmt` or `End_Select_Stmt`. Of course this "should not" happen but you never know...

Also, now I look at this again, is checking that str(child) contains 'DEFAULT' robust enough. e.g. do you need to worry about upper/lowercase and what if someone had a parameter named `default` and thus did `case(default)`?

Reply:
I am not sure about your first point, we are searching select_case_stmt, case_stmt and end_slelect_stmt to record the indices and selector name, everything else (like the case body statements) should just pass to the next iteration without raising an error.

Your second concern is right, I will convert the str(child) to lower now.

Reply:
Ah, I see - I'd not fully understood the code. Thanks for adding the case insensitivity. Is this now robust to someone having a variable named 'default' and using that in a `case` statement, e.g.:

    integer, parameter :: default = 4
    ...
    select case(my_var)
        case(0)
            call do_this()
        case(default)
            call do_that()
        default
            call error()
    end select

Reply:
The default alone is not syntactically correct and produces a compile-time error (or fparser2 syntax error).  I tried the other cases `case (default)`, `case (DEFAULT)`, `case default`, `case DEFAULT` and all seem to work with the current implementation.

Reply:
Last comment is wrong, the parenthesis are meaningful and `CASE (default)` should be comparing the selector with a symbol named 'default'.

Reply:
I fixed it by using fparser2 child type comparisons instead of string comparisons, it should work properly and be more robust now.